jan2025/helpers.js

```diff
@@ -0,0 +1,369 @@
+        );
+      }
+    });
+  }
+
+  if (Object.keys(passportData).includes("_constraints")) {
+    // _constraints is a list of two requests: planning data first, os roads second
+    const planningDataConstraints = passportData["_constraints"][0]["constraints"];
+    Object.entries(planningDataConstraints).forEach(([currentKey, currentData]) => {
+      if (Object.keys(combinedConstraintChanges).includes(currentKey)) {
+        const proposedKey = combinedConstraintChanges[currentKey];
+        currentData["fn"] = proposedKey;
+        newsessionData["passport"]["data"]["_constraints"][0]["constraints"][proposedKey] = currentData;
+        delete newsessionData["passport"]["data"]["_constraints"][0]["constraints"][currentKey];
+
+        Object.entries(breadcrumbs).forEach(([nodeId, crumb]) => {
+          if (crumb.data?.hasOwnProperty("_constraints")) {
+            newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["constraints"][proposedKey] = currentData;
+            delete newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["constraints"][currentKey];
+          }
+        });
+      }
+
+      if (currentKey.includes("article4")) {
+        const proposedKey = currentKey.replaceAll("article4", "articleFour");
+        currentData["fn"] = proposedKey;
+        newsessionData["passport"]["data"]["_constraints"][0]["constraints"][proposedKey] = currentData;
+        delete newsessionData["passport"]["data"]["_constraints"][0]["constraints"][currentKey];
+
+        Object.entries(breadcrumbs).forEach(([nodeId, crumb]) => {
+          if (crumb.data?.hasOwnProperty("_constraints")) {
+            newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["constraints"][proposedKey] = currentData;
+            delete newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["constraints"][currentKey];
+          }
+        });
+      }
+    });
+
+    const planningDataMetadata = passportData["_constraints"][0]["metadata"];
+    Object.entries(planningDataMetadata).forEach(([currentKey, currentData]) => {
+      // A4 is only relevant constraint in metadata
+      if (currentKey.includes("article4")) {
+        const proposedKey = currentKey.replaceAll("article4", "articleFour");
+        newsessionData["passport"]["data"]["_constraints"][0]["metadata"][proposedKey] = currentData;
+        delete newsessionData["passport"]["data"]["_constraints"][0]["metadata"][currentKey];
+
+        Object.entries(breadcrumbs).forEach(([nodeId, crumb]) => {
+          if (crumb.data?.hasOwnProperty("_constraints")) {
+            newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["metadata"][proposedKey] = currentData;
+            delete newsessionData["breadcrumbs"][nodeId]["data"]["_constraints"][0]["metadata"][currentKey];
+          }
+        });
+      }
+    });
+  }
+
+  return newsessionData;
+}
+
+const flagChanges = {
```

Comment:
Super handy to have these maps as artifacts as well as being required for functionality. The git history will outlive google sheets as well as remain static 👍